modules/ctrl_tcp/ctrl_tcp.c

```diff
@@ -0,0 +1,358 @@
+}
+
+
+static int encode_response(struct mbuf *resp, const char *token)
+{
+	struct re_printf pf = {print_handler, resp};
+	struct odict *od = NULL;
+	char *buf = NULL;
+	int err;
+
+	err = odict_alloc(&od, 8);
+	if (err)
+		return err;
+
+	resp->pos = NETSRING_HEADER_SIZE;
+	err = mbuf_strdup(resp, &buf, resp->end - NETSRING_HEADER_SIZE);
+	if (err)
+	{
+		mem_deref(od);
+		return err;
+	}
+
+	mbuf_reset(resp);
+	mbuf_init(resp);
+	resp->pos = NETSRING_HEADER_SIZE;
+
+	err |= odict_entry_add(od, "type", ODICT_STRING, "response");
+	err |= odict_entry_add(od, "response", ODICT_STRING, buf);
+
+	if (token)
+		err |= odict_entry_add(od, "token", ODICT_STRING, token);
+
+	if (err)
+		goto out;
+
+	err = json_encode_odict(&pf, od);
+	if (err)
+		warning("ctrl_tcp: failed to encode response json (%m)\n", err);
+
+ out:
+	mem_deref(buf);
+	mem_deref(od);
+
+	return err;
+}
+
+
+static bool command_handler(struct mbuf *mb, void *arg)
+{
+	struct ctrl_st *st = arg;
+	struct mbuf *resp = mbuf_alloc(2048);
+	struct re_printf pf = {print_handler, resp};
+	struct odict *od = NULL;
+	const struct odict_entry *oe_cmd, *oe_prm, *oe_tok;
+	char buf[256];
+	int err;
+
+	err = json_decode_odict(&od, 32, (const char*)mb->buf, mb->end, 16);
+	if (err) {
+		warning("ctrl_tcp: failed to decode JSON with %zu bytes (%m)\n",
+			mb->end, err);
+		goto out;
+	}
+
+	oe_cmd = odict_lookup(od, "command");
+	oe_prm = odict_lookup(od, "params");
+	oe_tok = odict_lookup(od, "token");
+	if (!oe_cmd) {
+		warning("ctrl_tcp: missing json entries\n");
+		goto out;
+	}
+
+	debug("ctrl_tcp: handle_command:  cmd='%s', params:'%s', token='%s'\n",
+	      oe_cmd ? oe_cmd->u.str : "",
+	      oe_prm ? oe_prm->u.str : "",
+	      oe_tok ? oe_tok->u.str : "");
+
+	re_snprintf(buf, sizeof(buf), "%s%s%s",
+		    oe_cmd->u.str,
+		    oe_prm ? " " : "",
+		    oe_prm ? oe_prm->u.str : "");
+
+	resp->pos = NETSRING_HEADER_SIZE;
+
+	/* Relay message to long commands */
+	err = cmd_process_long(baresip_commands(),
+			       buf,
+			       str_len(buf),
+			       &pf, NULL);
+	if (err) {
+		warning("ctrl_tcp: error processing command (%m)\n", err);
+	}
+
+	err = encode_response(resp, oe_tok ? oe_tok->u.str : NULL);
+	if (err) {
+		warning("ctrl_tcp: failed to encode response (%m)\n", err);
+		goto out;
+	}
+
+	resp->pos = NETSRING_HEADER_SIZE;
+	err = tcp_send(st->tc, resp);
+	if (err) {
+		warning("ctrl_tcp: failed to send the message (%m)\n", err);
+	}
+
+ out:
+	mem_deref(resp);
+	mem_deref(od);
+
+	return true;  /* always handled */
+}
+
+
+static void tcp_close_handler(int err, void *arg)
+{
+	struct ctrl_st *st = arg;
+
+	(void)err;
+
+	st->tc = mem_deref(st->tc);
+}
+
+
+static void tcp_conn_handler(const struct sa *peer, void *arg)
+{
+	struct ctrl_st *st = arg;
+
+	(void)peer;
+
+	/* only one connection allowed */
+	st->tc = mem_deref(st->tc);
+
+	(void)tcp_accept(&st->tc, st->ts, NULL, NULL, tcp_close_handler, st);
+	(void)netstring_insert(&st->ns, st->tc, 0, command_handler, st);
```

Comment:
is there a memory leak here ?

if the TCP connection is closed, and a new one is created, will the `st->ns` be derefed ?

Reply:
true. I'll fix it.